Utilities/mkpack.py

```diff
@@ -0,0 +1,307 @@
+            return [self.resfile]
+        
+        def data(self):
+            return load_resource_from_pbpack(self.resfile, self.resid)
+        
+        def sourcedesc(self):
+            return "resource ID %d from pack %s" % (self.resid, self.resfile)
+    
+    class ResourceFile(Resource):
+        def __init__(self, coll, j):
+            super(self.__class__, self).__init__(coll, j)
+            
+            self.file = "%s/%s" % (self.coll.root, j["input"]["file"], )
+        
+        def deps(self):
+            return [self.file]
+        
+        def data(self):
+            return load_resource_from_disk(self.file)
+        
+        def sourcedesc(self):
+            return self.file
+    
+    def __init__(self, fname, root = "."):
+        """
+        Load in a resource collection from a file, but don't load the
+        resources associated with it.  (That happens later.)
+        """
+        
+        self.jfname = fname
+        self.root = root
+        
+        with open(fname, 'r') as fd:
+            jdb = json.load(fd)
+        
+        if "references" in jdb:
+            self.references = {k: "%s/%s" % (self.root, v) for k, v in jdb["references"].iteritems()}
+        else:
+            self.references = {}
+        
+        self.resources = []
+        for res in jdb["resources"]:
+            if res["input"]["type"] == "file":
+                self.resources.append(ResourceCollection.ResourceFile(self, res))
+            elif res["input"]["type"] == "resource":
+                self.resources.append(ResourceCollection.ResourceRef(self, res))
+            else:
+                raise ValueError("unknown resource type %s" % (res["type"], ))
+    
+    def deps(self):
+        """
+        List of all dependencies of this resource pack.
+        """
+        
+        deps = {}
+        for r in self.resources:
+            for d in r.deps():
+                deps[d] = True
+        return [d for d in deps]
+    
+    def rsrcs(self):
+        """
+        List of raw resource data in this resource pack.
+        """
+        
+        return [r.data() for r in self.resources]
+    
+    def write_pbpack(self, fname):
+        """
+        Write out a .pbpack file with all of the resources loaded.
+        """
+        
+        return save_pbpack(fname, self.rsrcs())
+    
+    def write_header(self, fname):
+        """
+        Write out a C header file that has symbolic names for resource
+        identifiers.
+        """
+        
+        with open(fname, 'w') as fd:
+            fd.write("/* THIS FILE IS AUTOMATICALLY GENERATED BY mkpack.py. */\n")
+            fd.write("/* IF YOU MODIFY IT, YOU WILL BE VERY SAD. */\n")
+            fd.write("/* IF YOU CHECK IT IN, I WILL BE VERY SAD. */\n")
+            fd.write("\n")
+            fd.write("#pragma once\n")
+            fd.write("\n")
+            fd.write("typedef enum resource_id {\n")
+            for (id, r) in enumerate(self.resources):
```

Comment:
Don't use id as your variable name, it's reserved keyword(built-in function name). Use something like id_

Reply:
```
>>> help(id)
Help on built-in function id in module __builtin__:

id(...)
    id(object) -> integer

    Return the identity of an object.  This is guaranteed to be unique among
    simultaneously existing objects.  (Hint: it's the object's memory address.)
```

TIL.  Renamed to `rid`.